index.bs

```diff
@@ -52,10 +69,175 @@ dictionary StorageBucketOptions {
+
+1. If |bucket| is non-null and |bucket|'s [=StorageBucket/expiration value=] is non-null and less than or equal the [=current monotonic time=], then:
+
+    1. Remove |bucket|.
+
+    1. Set |bucket| to null.
+
+1. If |bucket| is null, then:
+
+    1. Let |bucket| be a new [=/storage bucket=] with name |name|.
+
+    1. Set |bucket|'s [=StorageBucket/durability value=] to |options|["{{StorageBucketOptions/durability}}"] if it exists.
+
+    1. Set |bucket|'s [=StorageBucket/quota value=] to |quota|.
+
+1. If |persisted| is true, set |bucket|'s [=/bucket mode=] to `"persistent"`.
+
+1. Set |bucket|'s [=StorageBucket/expiration value|expiration=] to |expires|.
+
+1. Let |storageBucket| be a new {{StorageBucket}}.
+
+1. Set |storageBucket|'s [=/storage bucket=] to |bucket|.
+
+1. Return |storageBucket|.
+
+</div>
+
+<div algorithm>
+
+To <dfn>validate a bucket name</dfn> given string |name|, run the following steps:
+
+1. If |name| contains any [=code point=] that is not [=ASCII lower alpha=], [=ASCII digit=], U+005F (_), or U+002D(-), then return failure.
+
+1. If |name| [=string/code point length=] is 0 or exceeds 64, then return failure.
+
+1. If |name| begins with U+005F (_) or U+002D(-), then return failure.
+
+1. Return.
+
+</div>
+
+When not null or undefined, {{StorageBucketOptions/durability}} is a hint to the user agent
+specifying the desired {{StorageBucket/durability}}. The user agent MAY
+create a new {{StorageBucket}} with this [=StorageBucket/durability value=]. The user agent
+MUST NOT modify the [=StorageBucket/durability value=] of an existing bucket.
+
 <h3 id="storage-bucket-delete">Deleting a bucket</h3>
+
+<div algorithm>
+
+The <dfn method for="StorageBucketManager">delete(|name|)</dfn> method steps are:
+
+1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+1. Let |shelf| be the result of running [=obtain a local storage shelf=] given |environment|.
+
+1. If |shelf| is failure, then [=exception/throw=] an "{{UnknownError}}" {{DOMException}} and abort these steps.
+
+1. Let |p| be [=a new promise=].
+
+1. Run the following steps [=in parallel=]:
+
+    1. If the result of [=validate a bucket name=] with |name| is failure, then [=/reject=] |p| with an {{InvalidCharacterError}} and abort these steps.
+
+    1. Let |bucket| be |shelf|'s [=bucket map=][|name|] if exists. Otherwise return.
+
+    1. Remove |bucket|.
+
+    1. [=/Resolve=] |p|.
+
+1. Return |p|.
+
+</div>
+
+Issue: [[Storage]] needs to define bucket removal. The bucket needs to be internally marked as
+removed and subsequent attempts to access it (through associated storage endpoints) should fail.
+
 <h3 id="storage-bucket-keys">Enumerating buckets</h3>
 
+<div algorithm>
+
+The <dfn method for="StorageBucketManager">keys()</dfn> method steps are:
+
+1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+1. Let |storageKey| be the result of running [=obtain a storage key=] given |environment|.
```

Comment:
Do you still need `|storageKey|` now that you are getting the shelf below?

Reply:
Removed